proposals/downloads_get_file_hash.md

```diff
@@ -0,0 +1,164 @@
+# Proposal: `downloads.getFileHash()`
+
+**Summary**
+
+Allow extensions to query hashes of contents of downloaded files
+programmatically.
+
+**Document Metadata**
+
+**Author:** @bershanskyi
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** @Rob--W
+
+**Created:** 2024-04-08
+
+**Related Issues:** [WECG GitHub issue 401](https://github.com/w3c/webextensions/issues/401)
+
+## Motivation
+
+### Objective
+
+This API will enable extensions to privately and efficiently scan user
+downloads for "Known Unsafe" contents and validate the integrity of "Known Safe"
+files without exposing the raw content of the files.
+
+#### Use Cases
+
+##### Filtering for unwanted downloads
+
+Many antivirus products detect previously seen "safe" and "unsafe" files by
+computing a file content hash and checking it against a set of known hashes.
+These checks require minimal effort on the client side, yet are powerful
+enough to limit the spread of known malware downloads since effectively every
+client gets "inoculated" against a particular download as soon as a single
+antivirus detects a malicious file.
+User-Agents may store file hashes even after the original file is deleted to
+facilitate tracing of malware exposure even if the file is deemed to be
+dangerous after the user initially downloads it.
+
+##### Download integrity assurance
+
+Files can get damaged while they are shared on a file hosting server or NAS.
+While some file formats explicitly support integrity checks like embedded
+checksums or even cryptographic signatures, other files do not. This API
+would allow file managers or extensions to programmatically check the
+integrity of downloaded files if the file source elects to provide them.
+
+### Known Consumers
+
+None at the moment.
+
+## Specification
+
+### Schema
+
+Schema in Chromium-style IDL:
+```idl
+namespace downloads {
+    callback GetFileHashCallback = void(DOMString hash);
+
+    [supportsPromises, permissions="downloads.getFileHash"]
+    static void getFileHash(
+        long downloadId,
+        [legalValues=("SHA-256")] DOMString algorithm,
+        GetFileHashCallback callback);
+}
+```
+
+Schema in TypeScript:
+```ts
+namespace downloads {
+  enum FileHashAlgorithm {
+    SHA256 = "SHA-256"
+  }
+
+  getFileHash: (
+    downloadId: number,
+    algorithm?: keyof downloads.FileHashAlgorithm)
+      => Promise<string>;
+}
+```
+
+### New Permissions
+
+|      Permission Added     | Suggested Warning |
+| ------------------------- | ----------------- |
+| `"downloads.getFileHash"` | No warning        |
+
+The new `"downloads.getFileHash"` permission is optional and is honored only if
+extension has `"downloads"` permission as well, and the extension has host
```

Comment:
@oliverdunk The discussion reminds me that not all URLs have an origin. E.g. data:-URLs. What is your recommendation for these cases?

In the case of data:-URLs specifically, since the data is already fully revealing its content, I don't think that any origin permissions are needed there.

Reply:
Good point! It looks like we do expose the full URL already there so I don't have any strong feelings. One thing that came up in discussion about this proposal is that we want to run it by the downloads API owners (which is not the general extensions team) and privacy / security review. I'm going to drive that but suspect this may be something where they will have more thoughts.

Reply:
@oliverdunk Did you have a chance to discuss this internally?

Reply:
I confirmed with the downloads API owners last week that they are comfortable with this change. I haven't yet been able to go through privacy/security review though, which I'll do as soon as possible.